File: ui-tests/PageObjects/BasePage.cs
using System.Threading.Tasks;
using Microsoft.Playwright;

namespace UtTestsExperimentalConsoleAppication.PageObjects;

/// <summary>
/// Base class for all pages in the Playwright test suite.
/// Holds a reference to the underlying <see cref="IPage"/>.
/// </summary>
public abstract class BasePage
{
    protected BasePage(IPage page) => Page = page;

    /// <summary>
    /// Playwright page driving the browser.
    /// </summary>
    protected IPage Page { get; }
}

/// <summary>
/// Representation of a generic tab within CodeTracer.
/// </summary>
public class TabObject
{
    private readonly string _tabButtonText;

    protected readonly IPage Page;

    /// <summary>
    /// Root locator of the tab content.
    /// </summary>
    public ILocator Root { get; }

    public TabObject(IPage page, ILocator root, string tabButtonText)
    {
        Page = page;
        Root = root;
        _tabButtonText = tabButtonText;
    }

    /// <summary>
    /// Locator for the tab button in the tab strip.
    /// </summary>
    public ILocator TabButton()
        => Page.Locator(".lm_title").Filter(new() { HasTextString = _tabButtonText }).First;

    /// <summary>
    /// Determines whether the tab is currently visible.
    /// </summary>
    public async Task<bool> IsVisibleAsync()
    {
        var style = await Root.Locator("..").GetAttributeAsync("style");
        return style is null || !style.Contains("none");
    }
}

File: tsc-ui-tests/page-objects/base-page.ts
import type { Locator, Page } from "@playwright/test";

/**
 * Base page abstraction. All page objects that need a Page reference
 * should extend this class.
 */
export class BasePage {
  readonly page: Page;

  constructor(page: Page) {
    this.page = page;
  }
}

/**
 * Representation of a generic pane/tab within the Golden Layout container.
 *
 * Provides the tab button locator and visibility check shared by all
 * specialized pane page objects.
 */
export class TabObject {
  readonly page: Page;
  readonly root: Locator;
  tabButtonText: string;

  constructor(page: Page, root: Locator, tabButtonText: string) {
    this.page = page;
    this.root = root;
    this.tabButtonText = tabButtonText;
  }

  tabButton(): Locator {
    return this.page
      .locator(".lm_title", { hasText: this.tabButtonText })
      .first();
  }

  async isVisible(): Promise<boolean> {
    const style = await this.root.locator("..").getAttribute("style");
    return !(style?.includes("none"));
  }
}
